numpy/lib/_format_impl.py

```diff
@@ -872,6 +872,9 @@ def read_array(fp, allow_pickle=False, pickle_kwargs=None, *,
                     data = _read_bytes(fp, read_size, "array data")
                     array[i:i + read_count] = numpy.frombuffer(data, dtype=dtype,
                                                              count=read_count)
+        
+        if array.size != count:
+            raise ValueError(f"Failed to read all data for array. Expected {shape} = {count} elements, read {array.size} elements.")
```

Comment:
Almost think an `IOError` would be nice, but let's stick with `ValueError`, since that was the old one.

I like this a lot more, thanks!  Thinking if we can polish it a bit, but maybe this is already it.  Maybe `"could only read ..."` or add something like: `(file seems not fully written)`  or so?  Thinking about when this happens, and I guess it is usually if for some reason the file wasn't written completely (disk space/error or just forgetting to flush if reading what another program just wrote).

---  

Mainly, the line is also too long.

@charris do you happen to know right away why this is not caught by the linter CI job?